import sys
import os
import logging
import tempfile
import time
from dotenv import load_dotenv
import uuid
import requests
import json

# åŠ è½½ç¯å¢ƒå˜é‡
load_dotenv()

# ç¡®ä¿èƒ½å¯¼å…¥ rpc ç›®å½•
sys.path.append(os.path.join(os.path.dirname(os.path.dirname(__file__)), 'rpc'))

# å…¨å±€åˆå§‹åŒ–ä¸€ä¸ª Session å¯¹è±¡
# å®ƒå¯ä»¥è‡ªåŠ¨ç»´æŒ TCP é•¿è¿æ¥ï¼Œé¿å…æ¯æ¬¡æ¡æ‰‹
http_session = requests.Session()
adapter = requests.adapters.HTTPAdapter(pool_connections=10, pool_maxsize=10)
http_session.mount('http://', adapter)

# ğŸ”¥ 1. å¯¼å…¥æ–°çš„ Proto å®šä¹‰
from rpc import rag_pb2
from rpc import rag_pb2_grpc

# å¼•å…¥æ ¸å¿ƒç»„ä»¶
from core.llm import LLMClient
from core.embedding import EmbeddingModel
from tools.pdf_parser import PDFParser

# å¼•å…¥å­˜å‚¨ç»„ä»¶ (MinIO & Qdrant)
from minio import Minio
from qdrant_client import QdrantClient
from qdrant_client.http import models as rest
from config import Config # å‡è®¾ä½ æŠŠé…ç½®éƒ½æ”¾åœ¨è¿™é‡Œäº†

# ç»§æ‰¿æ–°çš„ RagServiceServicer
class ChimeraLLMService(rag_pb2_grpc.RagServiceServicer):
    def __init__(self):
        logging.info("ğŸ› ï¸ åˆå§‹åŒ– Chimera RAG Service...")

        # 1. ç°æœ‰ç»„ä»¶
        self.llm = LLMClient()
        EmbeddingModel.get_instance() # é¢„åŠ è½½æ¨¡å‹

        # 2. ğŸ”¥ æ–°å¢ï¼šåˆå§‹åŒ– MinIO (ç”¨äºä¸‹è½½æ–‡ä»¶)
        # å»ºè®®åç»­å°è£…åˆ° core/storage.py
        self.minio = Minio(
            os.getenv("MINIO_ENDPOINT", "localhost:9000"),
            access_key=os.getenv("MINIO_AK", "chimera_minio"),
            secret_key=os.getenv("MINIO_SK", "chimera_minio_secret"),
            secure=False
        )

        # 3. ğŸ”¥ æ–°å¢ï¼šåˆå§‹åŒ– Qdrant (ç”¨äºå†™å…¥å‘é‡)
        # å»ºè®®åç»­å°è£…åˆ° core/vector_store.py
        self.qdrant = QdrantClient(
            host=os.getenv("QDRANT_HOST", "localhost"),
            port=int(os.getenv("QDRANT_PORT", 6333))
        )
        self.collection_name = "chimera_docs"

    # ----------------------------------------------------------------
    # 1. èŠå¤©æ¥å£ (ChatStream) - é€‚é… v0.4.0
    # ----------------------------------------------------------------
    def ChatStream(self, request, context):
        logging.info(f"[Chat] æ”¶åˆ°æé—®: {request.query} (KB={request.kb_id}, Org={request.org_id})")

        try:
            # ==========================================
            # Step 1: Embedding
            # ==========================================
            query_vector = EmbeddingModel.encode(request.query)
            if hasattr(query_vector, "tolist"):
                query_vector = query_vector.tolist()

            # ==========================================
            # Step 2: æ„å»º Filter (æ‰‹åŠ¨æ„å»ºå­—å…¸ï¼Œä¸ä¾èµ– SDK å¯¹è±¡)
            # ==========================================
            # Qdrant çš„ Filter JSON ç»“æ„
            filter_payload = None

            must_conditions = []
            if request.kb_id > 0:
                must_conditions.append({
                    "key": "kb_id",
                    "match": {"value": request.kb_id}
                })
            elif request.org_id > 0:
                must_conditions.append({
                    "key": "org_id",
                    "match": {"value": request.org_id}
                })

            if must_conditions:
                filter_payload = {"must": must_conditions}

            # ==========================================
            # Step 3: åŸç”Ÿ HTTP è¯·æ±‚ (æ ¸æ­¦å™¨çº§ä¿®å¤ â˜¢ï¸)
            # ==========================================
            # æ‹¼æ¥ Qdrant æœç´¢æ¥å£ URL
            qdrant_host = os.getenv("QDRANT_HOST", "localhost")
            qdrant_port = os.getenv("QDRANT_PORT", "6333")
            url = f"http://{qdrant_host}:{qdrant_port}/collections/{self.collection_name}/points/search"

            # æ„é€  Request Body
            payload = {
                "vector": query_vector,
                "limit": 15,
                "with_payload": True,
                "score_threshold": 0
            }
            if filter_payload:
                payload["filter"] = filter_payload

            logging.info(f"ğŸš€ å‘é€ HTTP æœç´¢è¯·æ±‚: {url}")

            # å‘é€è¯·æ±‚
            response = http_session.post(url, json=payload, timeout=5)

            if response.status_code != 200:
                logging.error(f"Qdrant HTTP Error: {response.text}")
                raise Exception(f"Search failed with status {response.status_code}")

            # è§£æç»“æœ
            # Qdrant HTTP æ¥å£è¿”å›æ ¼å¼: { "result": [ { "payload": {...}, "score": 0.9 }, ... ], ... }
            resp_json = response.json()
            search_results = resp_json.get("result", [])

            logging.info(f"ğŸ” æ£€ç´¢åˆ° {len(search_results)} æ¡ç›¸å…³ä¸Šä¸‹æ–‡")

            # ==========================================
            # Step 4: æ„å»º Context (è§£æåŸç”Ÿ JSON)
            # ==========================================
            if not search_results:
                context_str = "æ²¡æœ‰æ‰¾åˆ°ç›¸å…³çš„ä¸Šä¸‹æ–‡ä¿¡æ¯ã€‚"
            else:
                context_parts = []
                for idx, hit in enumerate(search_results):
                    # HTTP è¿”å›çš„ hit æ˜¯å­—å…¸
                    payload = hit.get("payload", {})

                    if payload:
                        content = payload.get("content", "")
                        file_name = payload.get("file_name", "unknown")
                        page_num = payload.get("page_number", 0)

                        part = f"ç‰‡æ®µ{idx+1}: {content}\næ¥æº: <<{file_name}|{page_num}>>"
                        context_parts.append(part)

                context_str = "\n\n".join(context_parts)

            # ==========================================
            # Step 5: è°ƒç”¨ LLM (ä¿æŒä¸å˜)
            # ==========================================
            final_system_prompt = f"""
            ä½ æ˜¯ä¸€ä¸ªä¸“ä¸šçš„ç§‘ç ”åŠ©æ‰‹ (Chimera-RAG)ã€‚
            è¯·åŸºäºä»¥ä¸‹ã€å‚è€ƒä¸Šä¸‹æ–‡ã€‘å›ç­”ç”¨æˆ·çš„ã€é—®é¢˜ã€‘ã€‚

            ã€å‚è€ƒä¸Šä¸‹æ–‡ã€‘
            {context_str}

            ã€å›ç­”è¦æ±‚ã€‘
            1. å¿…é¡»ä¸¥æ ¼å¼•ç”¨ä¸Šè¿°ä¸Šä¸‹æ–‡ä¸­çš„ä¿¡æ¯ã€‚
            2. å¦‚æœä¸Šä¸‹æ–‡æ²¡æœ‰æåˆ°ï¼Œè¯·ç›´æ¥è¯´ä¸çŸ¥é“ï¼Œä¸è¦ç¼–é€ ã€‚
            3. å¼•ç”¨æ ¼å¼ä¿æŒä¸ºï¼š<<æ–‡ä»¶å|é¡µç >>
            """

            generator = self.llm.stream_chat(request.query, system_prompt=final_system_prompt)

            for text_delta in generator:
                yield rag_pb2.ChatReply(answer_delta=text_delta)

        except Exception as e:
            import traceback
            traceback.print_exc()
            logging.error(f"âŒ RAG æµç¨‹å‡ºé”™: {e}")
            yield rag_pb2.ChatReply(answer_delta=f"**System Error**: {str(e)}")

    # ----------------------------------------------------------------
    # 2. æ–‡æ¡£å…¥åº“æ¥å£ (ParseAndIngest) - æ ¸å¿ƒé‡æ„
    # ----------------------------------------------------------------
    def ParseAndIngest(self, request, context):
        """
        ETL ç®¡é“ï¼šMinIOä¸‹è½½ -> Doclingè§£æ -> å‘é‡åŒ– -> Qdrantå…¥åº“
        """
        logging.info(f"\nğŸ“¥ [ETL] å¼€å§‹å¤„ç†ä»»åŠ¡: {request.file_name} (Path: {request.storage_path})")

        start_time = time.time()

        # ä½¿ç”¨ä¸´æ—¶æ–‡ä»¶å¤„ç†ï¼Œå¤„ç†å®Œè‡ªåŠ¨åˆ é™¤
        with tempfile.NamedTemporaryFile(suffix=".pdf", delete=True) as tmp_file:
            try:
                # --- Step 1: ä» MinIO ä¸‹è½½ ---
                logging.info(f"  â¬‡ï¸ æ­£åœ¨ä¸‹è½½: {os.getenv('MINIO_BUCKET', 'chimera-docs')}/{request.storage_path}")
                self.minio.fget_object(
                    os.getenv("MINIO_BUCKET", "chimera-docs"),
                    request.storage_path,
                    tmp_file.name
                )

                # ğŸ›¡ï¸ å®‰å…¨æ£€æŸ¥ï¼šç¡®è®¤æ–‡ä»¶çœŸçš„ä¸‹è½½ä¸‹æ¥äº†ä¸”ä¸ä¸ºç©º
                file_size = os.path.getsize(tmp_file.name)
                if file_size == 0:
                    raise Exception("MinIO ä¸‹è½½çš„æ–‡ä»¶ä¸ºç©ºï¼")
                logging.info(f"  âœ… ä¸‹è½½å®Œæˆï¼Œæ–‡ä»¶å¤§å°: {file_size} bytes")

                # --- Step 2: è°ƒç”¨ç°æœ‰çš„ PDFParser ---
                logging.info("  ğŸ“– è°ƒç”¨ Docling è§£æä¸­...")
                raw_chunks = PDFParser.parse_and_chunk(
                    file_source=tmp_file.name,
                    filename=request.file_name
                )

                if not raw_chunks:
                    return rag_pb2.ParseResponse(success=False, error_msg="è§£æç»“æœä¸ºç©º", chunk_count=0)

                # --- Step 3: å‘é‡åŒ– & å‡†å¤‡ Qdrant æ•°æ® ---
                points = []
                logging.info(f"  ğŸ§  æ­£åœ¨å‘é‡åŒ– {len(raw_chunks)} ä¸ªåˆ‡ç‰‡...")

                for idx, item in enumerate(raw_chunks):
                    content = item['content']
                    page_num = item.get('page', 1)

                    # è°ƒç”¨ Core Embedding
                    vector = EmbeddingModel.encode(content)

                    # æ„é€  Payload (å…ƒæ•°æ®)
                    payload = {
                        "content": content,
                        "file_name": request.file_name,
                        "page_number": page_num,
                        "doc_id": request.doc_id, # å…³è” Postgres ID
                        "kb_id": request.kb_id,   # çŸ¥è¯†åº“éš”ç¦»
                        "org_id": request.org_id  # ç»„ç»‡éš”ç¦»
                    }

                    # ä½¿ç”¨ uuid5 + DNS å‘½åç©ºé—´ï¼Œä¿è¯ "3_0" æ°¸è¿œè½¬æ¢æˆåŒä¸€ä¸ª UUID
                    unique_str = f"{request.doc_id}_{idx}"
                    point_id = str(uuid.uuid5(uuid.NAMESPACE_DNS, unique_str))

                    points.append(rest.PointStruct(
                        id=point_id,
                        vector=vector,
                        payload=payload
                    ))

                # --- Step 4: å†™å…¥ Qdrant ---
                logging.info(f"  ğŸ’¾ å†™å…¥ Qdrant ({len(points)} points)...")
                self.qdrant.upsert(
                    collection_name=self.collection_name,
                    points=points
                )

                duration = time.time() - start_time
                logging.info(f"âœ… [Success] ETL å®Œæˆï¼Œè€—æ—¶ {duration:.2f}s")

                return rag_pb2.ParseResponse(
                    success=True,
                    chunk_count=len(points),
                    page_count=0 # å¦‚æœ PDFParser è¿”å›äº†æ€»é¡µæ•°å¯å¡«è¿™é‡Œ
                )

            except Exception as e:
                import traceback
                traceback.print_exc()
                logging.error(f"âŒ ETL å¤±è´¥: {str(e)}")
                return rag_pb2.ParseResponse(
                    success=False,
                    error_msg=str(e),
                    chunk_count=0
                )